src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+
+#define GET_SCALE(_name) \
+    do { \
+        len = snprintf(path, sizeof(path), CHANNEL_SCALE_PATH, device->device_id, \
+            _name); \
+        if (len > 0 && sol_util_read_file(path, "%lf", &channel->scale) > 0) { \
+            goto end; \
+        } \
+    } while (0)
+
+    GET_SCALE(channel->name);
+    /* No scale. If channel has x, y, z or Y component, look for scale file without it */
+    channel_pure_name = channel_get_pure_name(channel);
+    GET_SCALE(channel_pure_name);
+
+    SOL_INF("Could not get scale for channel [%s] in device%d. Assuming 1.0",
+        channel->name, device->device_id);
+    channel->scale = 1.0;
+
+#undef GET_SCALE
+
+end:
+    free(channel_pure_name);
+}
+
+static void
+channel_get_offset(struct sol_iio_channel *channel)
+{
+    char path[PATH_MAX + 1];
+    int len;
+    char *channel_pure_name = NULL;
+    struct sol_iio_device *device = channel->device;
+
+#define GET_OFFSET(_name) \
+    do { \
+        len = snprintf(path, sizeof(path), CHANNEL_OFFSET_PATH, device->device_id, \
+            _name); \
+        if (len > 0 && sol_util_read_file(path, "%d", &channel->offset) > 0) { \
+            goto end; \
+        } \
+    } while (0)
+
+    GET_OFFSET(channel->name);
+    /* No scale. If channel has x, y or z component, look for scale file without it */
+    channel_pure_name = channel_get_pure_name(channel);
+    GET_OFFSET(channel_pure_name);
+
+    SOL_INF("Could not get offset for channel [%s] in device%d. Assuming 0",
+        channel->name, device->device_id);
+    channel->offset = 0;
+
+#undef GET_OFFSET
+
+end:
+    free(channel_pure_name);
+}
+
+static void
+iio_del_channel(struct sol_iio_channel *channel)
+{
+    free(channel->name);
+
+    free(channel);
+}
+
+void
+sol_iio_close(struct sol_iio_device *device)
+{
+    int i;
+    struct sol_iio_channel *channel;
+
+    SOL_PTR_VECTOR_FOREACH_REVERSE_IDX (&device->channels, channel, i) {
+        iio_del_channel(channel);
+        sol_ptr_vector_del(&device->channels, i);
+    }
+
+    if (device->fd_handler) sol_fd_del(device->fd_handler);
+    if (device->fd > -1) close(device->fd);
+    free(device->trigger_name);
+    free(device);
+}
+
+static bool
+read_channel_index_in_buffer(struct sol_iio_channel *channel)
+{
+    struct sol_iio_device *device = channel->device;
+    char path[PATH_MAX + 1];
+    int len;
+
+    len = snprintf(path, sizeof(path), CHANNEL_SCAN_INDEX_PATH,
+        device->device_id, channel->name);
+    if (len < 0 || sol_util_read_file(path, "%d", &channel->index) < 0)
+        return false;
+
+    return true;
+}
+
+struct sol_iio_channel *
+sol_iio_add_channel(struct sol_iio_device *device, const char *name)
+{
+    char path[PATH_MAX + 1];
+    struct sol_iio_channel *channel;
+    struct stat st;
+    int len;
+
+    len = snprintf(path, sizeof(path), CHANNEL_RAW_PATH, device->device_id,
+        name);
+    if (len < 0 || len > PATH_MAX || stat(path, &st)) {
```

Comment:
Check if `len >= PATH_MAX`, you have to account for the terminating NUL character.

Since you're calling stat here, what happens if, during program bootup the channel is there, but is removed due to a kernel module being removed? Will the program crash? It's often a bad idea to use stat() if you plan to open the file later (and assume a bunch of things if stat() or access() do not fail).